docs/root/overview.rst

```diff
@@ -0,0 +1,97 @@
+
+### Process
+
+**Process** represents the primary entry point to a Nighthawk execution run. Only one Process is allowed to exist at the same point in time within an OS Process. **Process **is responsible for performing process-wide initialization and termination, as well as handle input configuration, deliver output, and co-ordinate Workers. **ProcessImpl **is re-used accross the CLI and the gRPC service.
+
+### Worker
+
+**Worker **is responsible for performing correct initialization and termination of its thread, as well as execution of it’s designated task and offering a way for consumers to wait for that task to complete. 
+
+### TerminationPredicate
+
+**TerminationPredicate **is responsible for if and how to terminate. As of today, there are two types: one that will indicate that it is time to terminate based on a pre-configured duration, and one that will do so based on absolute counter thresholds. 
+
+### Sequencer
+
+**SequencerImpl** resides on a worker-thread, and drives itself via timers that run on the dispatcher, and coordinates interaction between **RateLimiter,** **BenchmarkClient**, and **TerminationPredicate** to drive execution to completion.
+
+### RateLimiter
+
+**RateLimiter **is responsible for indicating when it is time to release a request. **RateLimiter **offers a semaphore-like interaction model, as in closed-loop mode it may be that **BenchmarkClient **is not able to satisfy request-release timings, in which case acquisitions from **RateLimiter** need to be cancelled. Concretely, as of today there is **LinearRateLimiterImpl** which offers a straight-paced plain frequency, as well as work in progress on **DistributionSamplingRateLimiterImpl **(adding uniformly distributed random timing offsets to an underlying **RateLimiter**) and **RampingRateLimiter**. 
+
+### BenchmarkClient
+
+As of today, there’s a single implementation called** BenchmarkClientImpl,** which wraps Envoy’s **Upstream** concept and (slightly) customized H1/H2 **Pool** concepts. For executing requests, the pool will be requested to create a **StreamEncoder**, and Nighthawk will pass its own **StreamDecoderImpl** into that as an argument. The integration surface between **BenchmarkClient **is defined via **BenchmarkClient::tryStartRequest()** and a callback specification which will be fired upon completion of a successfully started request.
+
+For H3, it is anticipated that it will fit into this model, but if all else fails, it will be entirely possible to wire in a new type of **BenchmarkClient.**
+
+### RequestSource
+
+**RequestSource **is an abstraction that allows us to implement different ways for **BenchmarkClient** to get information on what the request that it is about to fire off should look like. Today, two implementations exist: a static one, which will repeat the same request over and over, as well as one that pulls dynamic request data from a grpc service. The latter can, for example, be used to implement log-replay.
```

Comment:
This is another place where we are referring to a "grpc service", can we replace it with a name of the service, since this is actually referring to a different service?

Reply:
So at this time we're mentioning something that has not landed yet; but when that does we could link to the service definition that one should implement to make use of this.. maybe we should consider updating this part of #137?

Reply:
Sounds good.